File: checker/src/middlewares/rateLimiter.ts
import { Context, Next } from "hono";
import { HTTPException } from "hono/http-exception";

interface RateLimitStore {
  [key: string]: {
    count: number;
    resetTime: number;
  };
}

export function createRateLimiter(options: {
  windowMs: number;
  max: number;
  message?: string;
  keyGenerator?: (c: Context) => string;
}) {
  const store: RateLimitStore = {};
  const { windowMs, max, message = "Too many requests, please try again later." } = options;
  const keyGen = options.keyGenerator || ((c) => {
    return c.req.header("x-forwarded-for") || c.req.header("x-real-ip") || "unknown-ip";
  });

  // Cleanup store periodically to avoid memory leaks
  const intervalId = setInterval(() => {
    const now = Date.now();
    for (const key in store) {
      if (store[key].resetTime < now) {
        delete store[key];
      }
    }
  }, windowMs * 2);

  // Unref the timer so it doesn't prevent the process from exiting in tests
  if (typeof intervalId.unref === "function") {
    intervalId.unref();
  }

  return async (c: Context, next: Next) => {
    const key = keyGen(c);
    const now = Date.now();
    
    if (!store[key] || store[key].resetTime < now) {
      store[key] = {
        count: 0,
        resetTime: now + windowMs,
      };
    }

    store[key].count++;

    c.header("X-RateLimit-Limit", String(max));
    c.header("X-RateLimit-Remaining", String(Math.max(0, max - store[key].count)));
    c.header("X-RateLimit-Reset", String(Math.ceil(store[key].resetTime / 1000)));

    if (store[key].count > max) {
      throw new HTTPException(429, { message });
    }

    await next();
  };
}

// In-memory login rate limiter: 5 attempts per 15 minutes
export const loginRateLimiter = createRateLimiter({
  windowMs: 15 * 60 * 1000,
  max: 5,
  message: "Too many login attempts from this IP, please try again after 15 minutes.",
  keyGenerator: (c) => `login-${c.req.header("x-forwarded-for") || c.req.header("x-real-ip") || "unknown-ip"}`
});

// Global checker rate limiter: 60 requests per minute
export const checkerRateLimiter = createRateLimiter({
  windowMs: 60 * 1000,
  max: 60,
  message: "Too many proxy check requests from this IP, please try again after a minute.",
  keyGenerator: (c) => `checker-${c.req.header("x-forwarded-for") || c.req.header("x-real-ip") || "unknown-ip"}`
});

File: microservice-checker/proxy-service/internal/shared/middleware/ratelimiter.go
package middleware

import (
	"fmt"
	"net/http"
	"sync"
	"time"

	"github.com/hidessh99/sub-cf-vpn/microservice-checker/proxy-service/internal/shared/dto"
	"github.com/labstack/echo/v4"
)

type rateLimitEntry struct {
	count     int
	resetTime time.Time
}

type RateLimiter struct {
	store     sync.Map
	window    time.Duration
	max       int
	message   string
	keyPrefix string
}

func NewRateLimiter(window time.Duration, max int, message string, keyPrefix string) *RateLimiter {
	rl := &RateLimiter{
		window:    window,
		max:       max,
		message:   message,
		keyPrefix: keyPrefix,
	}

	// Periodic cleanup worker in background to prevent memory leak
	go func() {
		ticker := time.NewTicker(window * 2)
		for range ticker.C {
			now := time.Now()
			rl.store.Range(func(key, value interface{}) bool {
				entry := value.(*rateLimitEntry)
				if entry.resetTime.Before(now) {
					rl.store.Delete(key)
				}
				return true
			})
		}
	}()

	return rl
}

func (rl *RateLimiter) Limit() echo.MiddlewareFunc {
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			ip := c.Request().Header.Get("X-Forwarded-For")
			if ip == "" {
				ip = c.Request().Header.Get("X-Real-IP")
			}
			if ip == "" {
				ip = c.RealIP()
			}

			key := fmt.Sprintf("%s-%s", rl.keyPrefix, ip)
			now := time.Now()

			var entry *rateLimitEntry
			val, loaded := rl.store.Load(key)
			if !loaded {
				entry = &rateLimitEntry{
					count:     0,
					resetTime: now.Add(rl.window),
				}
			} else {
				entry = val.(*rateLimitEntry)
				if entry.resetTime.Before(now) {
					entry = &rateLimitEntry{
						count:     0,
						resetTime: now.Add(rl.window),
					}
				}
			}

			entry.count++
			rl.store.Store(key, entry)

			res := c.Response()
			res.Header().Set("X-RateLimit-Limit", fmt.Sprintf("%d", rl.max))
			remaining := rl.max - entry.count
			if remaining < 0 {
				remaining = 0
			}
			res.Header().Set("X-RateLimit-Remaining", fmt.Sprintf("%d", remaining))
			res.Header().Set("X-RateLimit-Reset", fmt.Sprintf("%d", entry.resetTime.Unix()))

			if entry.count > rl.max {
				return c.JSON(http.StatusTooManyRequests, dto.APIResponse{
					Success: false,
					Message: rl.message,
					Error:   nil,
				})
			}

			return next(c)
		}
	}
}

// Global rate limiters matching TS definitions
var (
	LoginRateLimiter = NewRateLimiter(
		15*time.Minute,
		5,
		"Too many login attempts from this IP, please try again after 15 minutes.",
		"login",
	)

	CheckerRateLimiter = NewRateLimiter(
		60*time.Second,
		60,
		"Too many proxy check requests from this IP, please try again after a minute.",
		"checker",
	)
)
